[WGSL] Add a pass to rewrite globals

tadeuzagallo opened the pull request:
#### 5357882f7ca7dfe24290ebfd8611dca74bd46246
<pre>
[WGSL] Add a pass to rewrite globals
<a href="https://bugs.webkit.org/show_bug.cgi?id=251205">https://bugs.webkit.org/show_bug.cgi?id=251205</a>
&lt;rdar://problem/104691861&gt;

Reviewed by Myles C. Maxfield.

Add a new pass that removes global variables and creates structs compatible
with Metal argument buffers. For now, the pass only generates the structs based
on the globals available in the source program, but we should use the pipeline
layout in the future. Fow now what it does is:

1. Find all the globals in the source and remove them
2. Creates a struct per bind group to receive the globals
3. Traverse the call graph and identify which functions use which globals
4. Insert parameters in the functions that need to receive the globals
5. Rewrite the global accesses with struct accesses to load the globals from
   the structs received via the parameters.

In order for this patch to work correct, GlobalVariableRewriter needs to be updated
to use a ReferenceType, which is being introduce in PR #9174.

This PR also includes a fair amount of FIXMEs, but it was getting pretty large so I
paused here and will upload follow up patches shortly.

* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp: Added.
(WGSL::RewriteGlobalVariables::RewriteGlobalVariables):
(WGSL::RewriteGlobalVariables::run):
(WGSL::RewriteGlobalVariables::visit):
(WGSL::replace):
(WGSL::RewriteGlobalVariables::collectGlobals):
(WGSL::RewriteGlobalVariables::visitEntryPoint):
(WGSL::RewriteGlobalVariables::requiredGroups):
(WGSL::RewriteGlobalVariables::insertStructs):
(WGSL::RewriteGlobalVariables::insertParameters):
(WGSL::RewriteGlobalVariables::def):
(WGSL::RewriteGlobalVariables::read):
(WGSL::RewriteGlobalVariables::argumentBufferParameterName):
(WGSL::RewriteGlobalVariables::argumentBufferStructName):
(WGSL::rewriteGlobalVariables):
* Source/WebGPU/WGSL/GlobalVariableRewriter.h: Added.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::FunctionDefinitionWriter::visitArgumentBufferParameter):
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/259477@main">https://commits.webkit.org/259477@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c84c4a54fe63658947e4a820bf2026768068cf2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105005 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14085 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37896 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114269 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15221 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5007 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97324 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113282 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110762 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/94767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/26394 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7423 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/27753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7517 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4334 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13572 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/47305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6541 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9306 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
```diff
@@ -46,6 +47,12 @@ class IdentifierExpression final : public Expression {
     String m_identifier;
+
+protected:
+    // FIXME: This is necessary becasue we replace an IdentifierExpression with
```
nit: because

djg commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -0,0 +1,281 @@
+}
+
+
+auto RewriteGlobalVariables::requiredGroups() -> IndexSet
```
This style isn't consistent with the rest of the code. I tried it at one point and @litherum questioned it. I'm happy to use either syntax but we should be consistent.

litherum replied:
The purpose of the `->` is for namespacing.

`int Foo::myThingy()` is better than `auto myThingy -> int`

`Foo::Bar Foo::myThingy()` is worse than `auto Foo::myThingy() -> Bar`

litherum replied:
https://github.com/WebKit/WebKit/pull/9200

djg reviewed (commented):
Thanks for trying to keep the patches small and easily reviewable.

:shipit:

litherum commented on Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
```diff
@@ -46,6 +47,12 @@ class IdentifierExpression final : public Expression {
+    // FIXME: This is necessary becasue we replace an IdentifierExpression with
+    // a StructureAccess for globals, so the sizes must be compatible. We need to
+    // introduce an IdentityExpression and remove this hack
+    uint8_t m_padding[sizeof(StructureAccess) - sizeof(Expression) - sizeof(String)];
```
It sucks that we'd have to do this. Surely there's a better way...

I think it's worth trying hard to avoid using placement new to replace nodes with other nodes. More side-tables, perhaps?

tadeuzagallo replied:
I think if we use an identity node as I describe in the FIXME it's pretty safe to use placement new? One thing I was considering is moving the AST representation to be a little more structured as an IR, so that modifications can become easier. That would make a lot of things easier.

litherum commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -0,0 +1,281 @@
+    // FIXME: visit callee once we have any
+
+    for (auto& parameter : functionDecl.parameters())
+        def(parameter.name());
```
Usually we use full words

tadeuzagallo replied:
Will rename. We should probably also rename all the `[...]Decl` nodes.

litherum commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -0,0 +1,281 @@
+void replace(CurrentType& dst, Arguments&&... arguments)
+{
+    static_assert(sizeof(TargetType) <= sizeof(CurrentType));
+    new (&dst) TargetType(dst.span(), std::forward<Arguments>(arguments)...);
```
:sadpanda:

litherum commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -0,0 +1,281 @@
+
+void RewriteGlobalVariables::collectGlobals()
+{
+    auto& globalVars = m_callGraph.ast().globalVars();
```
Maybe this is nit-picking, but it seems counter-intuitive to me that the set of global variables would be a member of the AST. I'd expect the ownership model to be that both the AST and any derived info (like this global variable collection) would be siblings, owned by the same root object.

tadeuzagallo replied:
But globals in this context are source elements, so they must be part of the AST, no? Maybe this is a case of meaning different things by `global`? What is meant here is e.g. when we parse:

```
@group(x) @binding(y) var<uniform> u: U;
```

It has to end up in the AST.

litherum commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -0,0 +1,281 @@
+
+namespace WGSL {
+
+class RewriteGlobalVariables : public AST::Visitor {
```
I kind of think we should have a design of the compiler where there are 2 pieces:

1. A "frontend" which is WGSL-specific, and knows nothing about MSL. Its job is to do WGSL-specific tasks (e.g. type checking, l-value checking, etc.).
2. A "backend" which  knows about MSL. The backend is unable to fail; its whole job is to take the program which we've proved to be correct in the frontend, and to spit out MSL code for it.

And I think we should have a deliberate API between the two ends of the compiler, because that would dramatically help us if we ever want to add another backend generator language one day. (And, ideally all the code to the backend of the compiler would even be in a subdirectory named `msl/` or `Metal Shading Language/`, maybe even with a `MSL::` C++ namespace.)

E.g. This global variable rewriter is kind of an implementation detail of MSL; nothing in WGSL itself cares about global variables being in one form or another. So this would be in the MSL-specific portion. If we want to emit another language, we wouldn't need this pass, right?

tadeuzagallo replied:
I agree with you, but I don't really know yet how much of this would be useful for other backends, as you know my experience here is "limited". I do agree that at the moment this pass is tailored to MSL, as that's our only output, but the passes are constructed in a way in which they are easily composable, so it should be a matter of moving them and running them conditionally when the time comes. I'm still figuring some things out as I go, so I'm trying to keep things composable, but not worrying too much about the details, as long as it's something simple to refactor in the future. Does that make sense?

tadeuzagallo committed:
[WGSL] Add a pass to rewrite globals
https://bugs.webkit.org/show_bug.cgi?id=251205
<rdar://problem/104691861>

Reviewed by Myles C. Maxfield.

Add a new pass that removes global variables and creates structs compatible
with Metal argument buffers. For now, the pass only generates the structs based
on the globals available in the source program, but we should use the pipeline
layout in the future. Fow now what it does is:

1. Find all the globals in the source and remove them
2. Creates a struct per bind group to receive the globals
3. Traverse the call graph and identify which functions use which globals
4. Insert parameters in the functions that need to receive the globals
5. Rewrite the global accesses with struct accesses to load the globals from
   the structs received via the parameters.

In order for this patch to work correct, GlobalVariableRewriter needs to be updated
to use a ReferenceType, which is being introduce in PR #9174.

This PR also includes a fair amount of FIXMEs, but it was getting pretty large so I
paused here and will upload follow up patches shortly.

* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp: Added.
(WGSL::RewriteGlobalVariables::RewriteGlobalVariables):
(WGSL::RewriteGlobalVariables::run):
(WGSL::RewriteGlobalVariables::visit):
(WGSL::replace):
(WGSL::RewriteGlobalVariables::collectGlobals):
(WGSL::RewriteGlobalVariables::visitEntryPoint):
(WGSL::RewriteGlobalVariables::requiredGroups):
(WGSL::RewriteGlobalVariables::insertStructs):
(WGSL::RewriteGlobalVariables::insertParameters):
(WGSL::RewriteGlobalVariables::def):
(WGSL::RewriteGlobalVariables::read):
(WGSL::RewriteGlobalVariables::argumentBufferParameterName):
(WGSL::RewriteGlobalVariables::argumentBufferStructName):
(WGSL::rewriteGlobalVariables):
* Source/WebGPU/WGSL/GlobalVariableRewriter.h: Added.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::FunctionDefinitionWriter::visitArgumentBufferParameter):
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/259477@main
